### PR TITLE
Custom mapping types

### PR DESCRIPTION
**Pull request by MartinMystikJonas**
See https://github.com/contributte/apitte/issues/157

### CODE REVIEW

**Comment by MartinMystikJonas** [comment]
There are some CS and PHPStan errors but it is code from master so I left it unfixed so it wont mix with relevant changes

File: src/Core/Schema/Validation/RequestParameterValidation.php [hunk 1]
```diff
@@ -11,6 +11,13 @@
 class RequestParameterValidation implements IValidation
 {
 
+	private array $allowedTypes = EndpointParameter::TYPES;
+
+	public function __construct(array $allowedTypes = [])
```
Comment (f3l1x) [inline src/Core/Schema/Validation/RequestParameterValidation.php:16]: I'd put default value here -> EndpointParameter::TYPES.

File: src/Core/Mapping/RequestParameterMapping.php [hunk 2]
```diff
@@ -217,7 +217,7 @@ protected function normalize($value, EndpointParameter $parameter, ITypeMapper $
 		}
 	}
 
-	protected function getMapper(string $type): ?ITypeMapper
+	public function getMapper(string $type): ?ITypeMapper
```
Comment (f3l1x) [inline src/Core/Mapping/RequestParameterMapping.php:220]: If it's only for tests, use reflection.

Take a look at Liberator class (https://github.com/contributte/redis/blob/master/tests/cases/DI/RedisExtension.phpt#L136-L137).

**Review by f3l1x — COMMENTED** [review]
There should be some documentation update as well.

File: src/Core/Mapping/RequestParameterMapping.php (hunk 2 above)
Comment (MartinMystikJonas) [reply src/Core/Mapping/RequestParameterMapping.php:220]: It is for tests but it make sense for me to have this public.

File: src/Core/Schema/Validation/RequestParameterValidation.php (hunk 1 above)
Comment (MartinMystikJonas) [reply src/Core/Schema/Validation/RequestParameterValidation.php:16]: Ok

**Comment by MartinMystikJonas** [comment]
I will add docs

**Comment by MartinMystikJonas** [comment]
Updated. I added docs and also exception handling for custom type validation fails. I would keep RequestParameterMapping::getMapper public - it makes sense for it to be public.

**Comment by f3l1x** [comment]
Awesome, thanks.

**Pull request merged**